src/services/redisSessionStorage.ts

```diff
@@ -2,64 +2,187 @@
+import TokenService from "./tokenService";
+
+export interface ISessionState {
+  readonly expired: boolean;
+  readonly expireAt?: number;
+  readonly newToken?: string;
+  readonly user?: User;
+}
 
 export default class RedisSessionStorage implements ISessionStorage {
   constructor(
     private readonly redisClient: redis.RedisClient,
-    private readonly tokenDuration: number
+    private readonly tokenDurationSecs: number,
+    private readonly tokenService: TokenService
   ) {}
 
   /**
    * {@inheritDoc}
    */
-  public set(token: string, user: User): void {
-    // Set key to hold the string value. This data is set to expire (EX) after
-    // `this.tokenDuration` seconds.
-    // @see https://redis.io/commands/set
-    this.redisClient.set(token, JSON.stringify(user), "EX", this.tokenDuration);
+  public set(
+    token: string,
+    user: User,
+    timestamp: number
+  ): Promise<Either<Error, boolean>> {
+    return new Promise(resolve => {
+      // Set a key to hold the fields value.
+      // @see https://redis.io/commands/hmset
+      this.redisClient.hmset(
+        token,
+        {
+          data: JSON.stringify(user),
+          timestamp
+        },
+        (err, response) => {
+          if (err) {
+            return resolve(left<Error, boolean>(err));
+          }
+
+          resolve(right<Error, boolean>(response));
+        }
+      );
+    });
   }
 
   /**
    * {@inheritDoc}
    */
-  public get(token: string): Promise<Either<Error, User>> {
-    const client = this.redisClient;
+  public get(token: string): Promise<Either<Error, ISessionState>> {
+    return new Promise(resolve => {
+      // Get the fields for this key.
+      // @see https://redis.io/commands/hmget
+      this.redisClient.hgetall(token, (err, value) => {
+        if (err) {
+          // Client returns an error.
+          return resolve(left<Error, ISessionState>(new Error(err.message)));
+        }
+
+        if (value === null || value === undefined) {
+          // Null and undefined values are considered "not found".
+          return resolve(this.sessionNotFound());
+        }
 
+        // Check if the token has expired. We don't remove expired token
+        // because a client can later refresh the session.
+        const timestamp = +value.timestamp;
+        if (timestamp + this.tokenDurationSecs * 1000 < Date.now()) {
+          return resolve(
+            left<Error, ISessionState>(new Error("Token has expired"))
+          );
+        }
+
+        // We got a value, let's see if it's a valid User.
+        const errorOrUser = extractUserFromJson(value.data);
+
+        errorOrUser.fold(
+          // Not a valid User, return an error.
+          () => resolve(this.unableToDecodeUser()),
+
+          // We got a valid user from the session.
+          user =>
+            resolve(
+              right<Error, ISessionState>({
+                expireAt: timestamp + this.tokenDurationSecs * 1000,
+                expired: false,
```

Comment:
isn't `expired` redundant if you already have `expireAt` ?

Reply:
in an initial discussion with @fpersico we designed the return of this endpoint with both expireAt and expired. In effect expired is redundant.